src/microbe_stage/editor/upgrades/ChemoreceptorUpgradeGUI.cs

```diff
@@ -17,102 +29,156 @@ public class ChemoreceptorUpgradeGUI : VBoxContainer, IOrganelleUpgrader
 
-        foreach (var choice in shownChoices)
+        foreach (var choice in shownCompoundChoices)
         {
             compounds.AddItem(choice.Name);
         }
 
-        // Select glucose by default
-        var defaultCompoundIndex =
-            shownChoices.FindIndex(c => c.InternalName == Constants.CHEMORECEPTOR_DEFAULT_COMPOUND_NAME);
+        shownSpeciesChoices = currentGame.GameWorld.Map.FindAllSpeciesWithPopulation();
+        shownSpeciesChoices.Sort((x, y) => string.Compare(x.Genus, y.Genus, StringComparison.OrdinalIgnoreCase));
 
-        if (defaultCompoundIndex < 0)
-            defaultCompoundIndex = 0;
+        foreach (var choice in shownSpeciesChoices)
+        {
+            species.AddItem(choice.Genus + " " + choice.Epithet);
+        }
+
+        var defaultCompound = SimulationParameters.Instance.GetCompound(Constants.CHEMORECEPTOR_DEFAULT_COMPOUND_NAME);
+        ChemoreceptorUpgrades defaultConfiguration = new ChemoreceptorUpgrades(defaultCompound, null,
+            Constants.CHEMORECEPTOR_RANGE_DEFAULT, Constants.CHEMORECEPTOR_AMOUNT_DEFAULT, defaultCompound.Colour);
 
         // Apply current upgrade values or defaults
         if (organelle.Upgrades?.CustomUpgradeData is ChemoreceptorUpgrades configuration)
         {
-            compounds.Selected = shownChoices.FindIndex(c => c == configuration.TargetCompound);
-            maximumDistance.Value = configuration.SearchRange;
-            minimumAmount.Value = configuration.SearchAmount;
-            colour.Color = configuration.LineColour;
+            LoadConfiguration(configuration, shownCompoundChoices, shownSpeciesChoices);
         }
         else
         {
-            compounds.Selected = defaultCompoundIndex;
-            maximumDistance.Value = Constants.CHEMORECEPTOR_RANGE_DEFAULT;
-            minimumAmount.Value = Constants.CHEMORECEPTOR_AMOUNT_DEFAULT;
-            colour.Color = shownChoices[defaultCompoundIndex].Colour;
+            LoadConfiguration(defaultConfiguration, shownCompoundChoices, shownSpeciesChoices);
         }
     }
 
     public bool ApplyChanges(ICellEditorComponent editorComponent, OrganelleUpgrades organelleUpgrades)
     {
-        if (shownChoices == null)
+        if (shownCompoundChoices == null || shownSpeciesChoices == null)
         {
             GD.PrintErr("Chemoreceptor upgrade GUI was not opened properly");
             return false;
         }
 
-        // Force some compound to be selected
+        // Force some type/compound/species to be selected
+        if (targetTypes.Selected == -1)
+            targetTypes.Selected = 0;
         if (compounds.Selected == -1)
             compounds.Selected = 0;
+        if (species.Selected == -1)
+            species.Selected = 0;
+
+        // Only one type of object can be detected
+        Compound? choiceCompound = null;
+        Species? choiceSpecies = null;
+
+        if (targetTypes.Selected == 0)
```

Comment:
Seeing as 0 and 1 are used as magic numbers in multiple places, I think you should specify an enum like in other places where different tabs are used in the GUI to add some bulletproofing to the code.